### DISCUSSION TITLE
[Proposal] Use verbatim identifier to allow a class member to have the same name as the class

### DISCUSSION THREAD

**Discussion by murshex** [dotnet/csharplang#6468] [Language Ideas]
for example:

```csharp
public class ExampleSingleton
{
    public static ExampleSingleton @ExampleSingleton = new();
    
    // Omitted for brevity
}
```

What are your thoughts? Sometimes I want to have a singleton global variable, but I don't know what to name it. Right now I'm prefixing all of my global singletons with an underscore like this "_ExampleSingleton" (also considered all caps like EXAMPLE_SINGLETON), but realized the @ symbol might be more appropriate. It doesn't work, it says that class members cannot have the same name as their enclosing type. I think this should be changed.

**Comment by huoyaoyuan** [comment]
```csharp
public class @class
{

}
```

The class name itself may require verbatim identifier.

**Reply by murshex** [reply]
The class name isn't using a keyword, I want to be able to have a class member with the same name as the class.

**Reply by 333fred** [reply]
Right, but how do you propose this feature works for a field named class inside a class named class?

**Reply by murshex** [reply]
Is there a real use case where someone wants to name a class "class"?
It doesn't even follow the naming convention of PascalCase too...
but I'll admit I didn't think about this situation...

after thinking about it,
I think not being able to have a class member have the same name as the class if the class name is already using the verbatim identifier is an acceptable compromise

**Reply by HaloFour** [reply]
```cs
public class @class {
    public @class @@class => this;
}
```

**Reply by murshex** [reply]
Sure, double `@` symbol is also a possibility, but is it even worth it?

**Comment by ufcpp** [comment]
Your idea seems odd to me because `A` and `@A` means exactly the same identifier in C#. The following code is likely to cause confusion.

```cs
class A
{
    public static B @A = new();
    public static void M() => Console.WriteLine("A static M");

    public static void Main()
    {
        A.M(); // A static M? or B instance M?
    }
}

class B
{
    public void M() => Console.WriteLine("B instance M");
}
```

**Reply by murshex** [reply]
Well, according to intellisense, `A` and `@A` means exactly the same identifier.
but heres the thing, having a class member with the same name as the class doesn't work at all to begin with.

And lets say I have a variable named `@namespace`, I can't just start using this variable without the `@` symbol. because thats a reserved keyword.

The same thing applies to the class member `@A`. I can't just start using this variable without the `@` symbol, because thats the name of the class.

I'm proposing that we allow class members with the same name as the class to be possible, like so:

`A.M()` would call the method statically,
while `@A.M()` would call the method on the instance of B class.

P.S. I'm using your library ContextFreeTask in my project :) I really love it!

**Reply by ufcpp** [reply]
`@namespace` (or `@class` and so on) is syntax for escaping reserved keywords, while your proposed syntax is for choosing one from the identifiers. The difference between the two is too large to use the same grammar.

P.S. thank you for using my ContextFreeTask!

**Comment by svick** [comment]
> Sometimes I want to have a singleton global variable, but I don't know what to name it.

I think `Instance` (or something similar) is a good name for such a member. Especially since something like `ReferenceEqualityComparer.Instance` reads well as "instance of `ReferenceEqualityComparer`", whereas `ReferenceEqualityComparer.ReferenceEqualityComparer` does not make sense to me.

**Reply by murshex** [reply]
That’s true but with C# 10 and `global using`, we can do 

`global using static ReferenceEqualityComparer;` and now use `@ReferenceEqualityComparer`  globally which is more readable because it’s less verbose than using `ReferenceEqualityComparer.Instance` everywhere…

I know some people think singletons are an anti pattern… but when you don’t want to use an IoC framework maybe because of personal preferences or hardware constraints, singletons are the next best thing as far as I’m concerned.

**Reply by svick** [reply]
In that case, I would probably reverse the pattern and have `Instances.ReferenceEqualityComparer`. The `Instances` class would be shared by all singletons in a project and could be declared as `partial`. Also, it would be usable both with and without `using static` (and you would actually need fewer `using static` lines than before).

**Reply by murshex** [reply]
Yes I agree with you. but one problem with this,

Let’s assume `ReferenceEqualityComparer` is in a namespace called `TestNamespace`. if I have `global using static Instances;` and then I want to use ReferenceEqualityComparer in a namespace called `TestNamespace.Example`, it’s not going to work because it thinks I’m trying to access the class type with the name `ReferenceEqualityComparer` instead of the static variable inside of the `Instances` class.

**Reply by CyrusNajmabadi** [reply]
> global using static ReferenceEqualityComparer; and now use @ReferenceEqualityComparer globally which is more readable because it’s less verbose than using 

This would be more confusing for me as this code is already legal and would mean a reference to a type. E.g. you can already place an @ on any identifier today and it has meaning.

**Reply by CyrusNajmabadi** [reply]
This also is not compelling for me as RefEqualityComparer.Instance is already completely clear and simple. It doesn't feel like a good thing to me to make that briefer and more confusing

**Reply by murshex** [reply]
Example:
```csharp
global using static Globals.Instances;
using TestNamespace;
using System;

namespace Globals
{
    public static class Instances
    {
        public static ReferenceEqualityComparer ReferenceEqualityComparer = new();
    }
}

namespace TestNamespace
{
    public class ReferenceEqualityComparer
    {
        public int GetData() => 420;
    }
}

namespace TestNamespace.Example
{
    public class ExampleClass
    {
        public void TestFunc()
        {
            // Error: CS0120: An object reference is required for the non-static field, method, or property 'ReferenceEqualityComparer.GetData()'
            var data = ReferenceEqualityComparer.GetData();
            Console.WriteLine($"data; {data}");
        }
    }
}
```

**Reply by murshex** [reply]
@CyrusNajmabadi IMHO when having over 35+ singletons, it really starts to get repetitive and is an eye sore having `SingletonClass.Instance` everywhere.

**Reply by murshex** [reply]
@CyrusNajmabadi What is the usefulness of being able to put @ on any class type if its not a keyword?

**Reply by HaloFour** [reply]
I prefer `Instance`, makes it very clear what that is.  Otherwise it's difficult to discern if you're accessing static members of a class or instance members of a property.

**Reply by murshex** [reply]
I mean, they added `global using static` for a reason, so might as well make use of it. From what I've noticed, global variable conventions in most other programming languages don't explicitly write `SingletonClass.Instance` everytime.

To be clear, I think using `SingletonClass.Instance` is fine for a publicly distributed library. But if I'm working on a console app, taking advantage of `global using static` to avoid having to write `SingletonClass.Instance` everywhere is worth it.

**Reply by HaloFour** [reply]
Singleton patterns in many OOP languages do typically involve an `instance` member.  This was not why global statics were added, your friction comes very specifically from wanting the member to match the type, but there's no requirement to do that and names are cheap, so you can already solve the problem by using a different name.  Reworking identifier resolution specifically for this use case, which I would argue should not be encouraged, does not feel worth it.

**Reply by murshex** [reply]
Why not. Think about it. Singletons are guaranteed to have one class instance. Majority of the time, I'm going to be calling `SingletonClass.Instance` to get access to the class instance. Only in very rare/limited circumstances will I actually ever have to use the class type `SingletonClass`. I really wish I could just write `SingletonClass` to get access to the class instance. I'd much rather explicity call something like `SingletonClass.ClassType` everytime I needed to use the class type, instead of explicitly calling the class instance everytime when thats all I need 99% of the time.

**Reply by CyrusNajmabadi** [reply]
> What is the usefulness of being able to put @ on any class type if its not a keyword?

It's the reverse. It was just always allowed. Note: a primary case is generators (existing, or the new source-generator variety). The generator can always use @ and is then fine whether the identifier is a keyword or not. Useful when taking apis from other domains and projecting into c#.

**Reply by murshex** [reply]
@CyrusNajmabadi Okay that I understand...

**Reply by CyrusNajmabadi** [reply]
> I mean, they added global using static for a reason,

We really added 'global using' and 'using static'. The combination just fell out. It's not something we necessarily think it's a good thing to do :-)

> so might as well make use of it. 

That's definitely not a position I ascribe to. I would use things judiciously, not just because you can. Thanks!

**Reply by CyrusNajmabadi** [reply]
> Why not. Think about it. Singletons are guaranteed to have one class instance. Majority of the time, I'm going to be calling SingletonClass.Instance

I mean, I get it. But calling the instance `SingletonClassInstance` seems totally fine to me. Easy to name, easy to consume (with static using or not) :-)

**Reply by murshex** [reply]
@CyrusNajmabadi I meant, might as well make use of it... if its useful. I really would hope that the csharp language maintainers would be more careful and not just let random language features appear because "the combination just fell out". If they really felt strongly that `global using static` does not have applicable use cases, I'm sure it wouldn't be a feature of the language...

**Reply by HaloFour** [reply]
I'd much sooner rather C# adopt companion objects, to be honest.

I also find singletons and globals to generally be a "bad thing".  I don't frequently make use of the pattern, I much rather components to not be aware of the lifecycle of its dependencies.  Just because a combination of existing features can be used to accomplish a thing like emulating globals doesn't mean it should, or that the language should be changed to encourage it further.

**Reply by murshex** [reply]
Singletons are probably used in 99% of applications, so I don't know how you can say they are bad haha

**Reply by HaloFour** [reply]
For extremely simple horizontal cases like a comparer, sure.  Those can be tested in isolation and you rarely expect to have to test the interactions between any class that depends on it.  But for any business logic I firmly believe that any given component should not be aware of the specific implementations of any of its dependencies or how to obtain them, nor should they make any assumptions regarding their lifecycle.  Doing such would only complicate refactoring if an existing component needs to take a dependency on another component with a different lifecycle, not to mention complicate the testing of those individual components.

**Reply by CyrusNajmabadi** [reply]
> not just let random language features appea

It didn't randomly appear. There were using statements, which became global.  Using them judiciously is the expectation. 

I don't expect people to use 'using static' where not appropriate, even though it's there. The same is true for the global form. Use it where it makes sense.

**Reply by murshex** [reply]
@CyrusNajmabadi Can you give an example of a good use case for global using static ?

**Reply by CyrusNajmabadi** [reply]
Any place where you would use a using-static, but would have that in all your files. Same as the cases for any other global-using :-)

**Reply by murshex** [reply]
Ok but using static is not ok to use for singletons ?

**Reply by CyrusNajmabadi** [reply]
I don't see any problem with singletons used appropriately. I just don't see why it's necessary or desirable to refer to them with the name of their containing type. That's the part that hasn't been an issue to me since c#1.  Singletons have been around since then, and giving them reasonable names hasn't been an issue afaict :-)

**Reply by CyrusNajmabadi** [reply]
> Ok but using static is not ok to use for singletons ?

If it makes your code more confusing, then probably not. If it makes your code clearer, then sure, use them :-) 

Off you want to use them for that purpose, I would name the Singleton well to make it clear what is going on. E.g. Foo.FooInstance. Now it's clear what is meant when you see the single identifier on its own :-)

**Comment by CyrusNajmabadi** [comment]
> does not have applicable use cases, I'm sure it wouldn't be a feature of the language...

I didn't say it didn't have applicable use cases. Just that not all use cases are compelling ones.  The cases you have specified are not compelling to me, that's all :-)
